docs: delete / cancel query (#11708)
apache/druid: docs/querying/sql.md

### HTTP POST

To use the SQL API to make Druid SQL queries, POST your query to the following endpoint on either the Router or Broker:
```

> POST https://ROUTER:8082/druid/v2/sql/`. 

Reviewer: Router uses the port 8888 by default while broker uses 8082.

through a JSON parsing error or through a missing trailing newline, you should assume the response was not fully
delivered due to an error.

### HTTP DELETE
You can use the HTTP `DELETE` method to cancel a SQL query using the following syntax:

> `DELETE https://BROKER:8082/druid/v2/sql/{sqlQueryId}`. The DELETE method requires the `sqlQueryId`, therefore to predict the query id you must set it in the query context.

Author: @jihoonson do we want to say something about what happens behind the scenes with this query cancellation?

through a JSON parsing error or through a missing trailing newline, you should assume the response was not fully
delivered due to an error.

### HTTP DELETE

> You can use the HTTP `DELETE` method to cancel a SQL query. When you cancel a query canceled, Druid marks its `SqlLifecycle` canceled immediately. After that, any operations in `SqlLifecycle` become non-operational to avoid unnecessary execution of expensive operations.

Reviewer 2: ```suggestion
You can use the HTTP `DELETE` method to cancel a SQL query. When you cancel a query, Druid immediately marks its `SqlLifecycle` as canceled. After that, any operations in `SqlLifecycle` become non-operational to avoid unnecessary execution of expensive operations.
```

> HTTP DELETE method uses the following syntax:

Reviewer 2: ```suggestion
Druid SQL's HTTP DELETE method uses the following syntax:
```

```

> DELETE https://BROKER:8082/druid/v2/sql/{sqlQueryId}

Reviewer: I'm wondering whether we should promote using router instead of broker since the router is now sort of mandatory to use the web console. Or, should we mention that you can use either router or broker address here?

Author: The POST examples used BROKER, but I can. change to ROUTER and mention both are options.

> DELETE https://BROKER:8082/druid/v2/sql/{sqlQueryId}

Reviewer 2: I like that this is clearly laid out in a code block, but it's a different format than is used to describe `HTTP POST`, which says: "You can make Druid SQL queries using HTTP via POST to the endpoint `/druid/v2/sql/`." 

Perhaps have a DELETE endpoint description similar to HTTP POST, or edit HTTP POST to have a code block like done here?

```

> The DELETE method requires the `sqlQueryId`, therefore to predict the query id you must set it in the query context.

Reviewer 2: ```suggestion
The DELETE method requires the `sqlQueryId` path parameter. To predict the query id you must set it in the query context.
```

> The DELETE method requires the `sqlQueryId`, therefore to predict the query id you must set it in the query context.

Author: I was wondering about that actually. Thank you!

For example if you issue the following query:
```bash
curl --request POST 'https://BROKER:8082/druid/v2/sql' \
--header 'Content-Type: application/json' \
--data-raw '{"query" : "SELECT sleep(CASE WHEN sum_added > 0 THEN 1 ELSE 0 END) FROM wikiticker WHERE sum_added > 0 LIMIT 15",
"context" : {"sqlQueryId" : "myQuery01"}}'
```
You can cancel the query using the query id as follows:
```bash
curl --request DELETE 'https://BROKER:8082/druid/v2/sql/myQuery01' \
```

> Druid returns an HTTP 202 response for successful deletion requests. If the the query id is incorrect or if the query completes before your cancellation request, Druid returns an HTTP 404 response. 

Reviewer: The cancellation request requires the read permission on all resources (currently datasource and view) in the sql query. It returns 404 as well when authorization fails.

Reviewer 3: Hi @jihoonson Why is status code 404 returned for autorization failure? In compliance with the HTTP status code definition,  code 403 is the appropriate one. Actually, the native query endpoint returns 403 for such case

https://github.com/apache/druid/blob/7220d0466bc31dda4b5396742ad08e18f9e01250/server/src/main/java/org/apache/druid/server/QueryResource.java#L164-L166

Reviewer: Per the RFC for HTTP, you can return either 403 or 404 (https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.3, https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.4). From security perspective, 404 is preferred to 403 since 403 leaks the existence of resource.

Reviewer 3: But in such case, this endpoint is already public to everyone, I think there's no concern of security problem so that we need to hide its existence. 404 might be a confusion for users when authorization fails. And it's better that we keep it consistent with the implementation of native query.

Reviewer: The SQL authorization checks the permission on the resources used in the SQL query, such as datasources or views. Returning 404 is to hide the existence of those datasources or views, not the endpoint. For user confusion, what is the use case where the user should know that he doesn't have permission to cancel the query? The query is usually cancelled by the user who submitted the query or the cluster admin. Either case, they will have the right permission already. For the difference from native query behavior, I'm pretty sure that most clients should handle native queries and sql queries separately anyway since their result format is different.

BUT, I think authorization failure handling should be consistent with the query posting endpoint which returns 403. I will create a PR to fix this. Thank you for bringing this up!

Reviewer: Raised https://github.com/apache/druid/pull/11710.

Reviewer 3: Thank you @jihoonson

delivered due to an error.

### HTTP DELETE
You can use the HTTP `DELETE` method to cancel a SQL query on either the Router or the Broker. When you cancel a query, Druid handles the cancellation in a best-effort manner. It marks the query canceled immediately and aborts the query execution as soon as possible. However, your query may run for a short time after your cancellation request.

Druid SQL's HTTP DELETE method uses the following syntax:
```
DELETE https://ROUTER:8082/druid/v2/sql/{sqlQueryId}
```

> The DELETE method requires the `sqlQueryId` path parameter. To predict the query id you must set it in the query context. Druid does not enforce unique `sqlQeryId` in the query context. If you issue a cancel request for a `sqlQeryId` active in more than one query context, Druid cancels all requests that use the query id.

Reviewer: ```suggestion
The DELETE method requires the `sqlQueryId` path parameter. To predict the query id you must set it in the query context. Druid does not enforce unique `sqlQueryId` in the query context. If you issue a cancel request for a `sqlQueryId` active in more than one query context, Druid cancels all requests that use the query id.
```

For example if you issue the following query:
```bash
curl --request POST 'https://ROUTER:8082/druid/v2/sql' \
--header 'Content-Type: application/json' \
--data-raw '{"query" : "SELECT sleep(CASE WHEN sum_added > 0 THEN 1 ELSE 0 END) FROM wikiticker WHERE sum_added > 0 LIMIT 15",
"context" : {"sqlQueryId" : "myQuery01"}}'
```
You can cancel the query using the query id `myQuery01` as follows:
```bash
curl --request DELETE 'https://ROUTER:8082/druid/v2/sql/myQuery01' \
```

Cancellation requests require READ permission on all resources used in the sql query. 

Druid returns an HTTP 202 response for successful deletion requests.

Druid returns an HTTP 404 response in the following cases:
  - Authorization fails.

>   - The query id is incorrect.

Reviewer: Since Druid uses `queryId` for native queries, perhaps it would be nice to make it clear that we are talking about `sqlQueryId`.

```suggestion
  - `sqlQueryId` is incorrect.
```

Druid returns an HTTP 404 response in the following cases:

>   - Authorization fails.

Reviewer: Sorry @techdocsmith, now it return 403. Can you update the doc one last time?
